vhost-device-vsock/src/thread_backend.rs

```diff
@@ -49,17 +54,144 @@ impl RawVsockPacket {
+            StreamType::Vsock(stream) => stream.flush(),
+        }
+    }
+}
+
+impl AsRawFd for StreamType {
+    fn as_raw_fd(&self) -> RawFd {
+        match self {
+            StreamType::Unix(stream) => stream.as_raw_fd(),
+            StreamType::Vsock(stream) => stream.as_raw_fd(),
+        }
+    }
+}
+
+impl ReadVolatile for StreamType {
+    fn read_volatile<B: BitmapSlice>(
+        &mut self,
+        buf: &mut VolatileSlice<'_, B>,
+    ) -> StdResult<usize, VolatileMemoryError> {
+        match self {
+            StreamType::Unix(stream) => stream.read_volatile(buf),
+            // Copied from vm_memory crate's ReadVolatile implementation for UnixStream
+            StreamType::Vsock(stream) => {
+                let fd = stream.as_raw_fd();
+                let guard = buf.ptr_guard_mut();
+
+                let dst = guard.as_ptr().cast::<libc::c_void>();
+
+                // SAFETY: We got a valid file descriptor from `AsRawFd`. The memory pointed to by `dst` is
+                // valid for writes of length `buf.len() by the invariants upheld by the constructor
+                // of `VolatileSlice`.
+                let bytes_read = unsafe { libc::read(fd, dst, buf.len()) };
+
+                if bytes_read < 0 {
+                    // We don't know if a partial read might have happened, so mark everything as dirty
+                    buf.bitmap().mark_dirty(0, buf.len());
+
+                    Err(VolatileMemoryError::IOError(std::io::Error::last_os_error()))
+                } else {
+                    let bytes_read = bytes_read.try_into().unwrap();
+                    buf.bitmap().mark_dirty(0, bytes_read);
+                    Ok(bytes_read)
+                }
+            }
+        }
+    }
+}
+
+impl WriteVolatile for StreamType {
+    fn write_volatile<B: BitmapSlice>(
+        &mut self,
+        buf: &VolatileSlice<'_, B>,
+    ) -> StdResult<usize, VolatileMemoryError> {
+        match self {
+            StreamType::Unix(stream) => stream.write_volatile(buf),
+            // Copied from vm_memory crate's WriteVolatile implementation for UnixStream
+            StreamType::Vsock(stream) => {
+                let fd = stream.as_raw_fd();
+                let guard = buf.ptr_guard();
+
+                let src = guard.as_ptr().cast::<libc::c_void>();
+
+                // SAFETY: We got a valid file descriptor from `AsRawFd`. The memory pointed to by `src` is
+                // valid for reads of length `buf.len() by the invariants upheld by the constructor
+                // of `VolatileSlice`.
+                let bytes_written = unsafe { libc::write(fd, src, buf.len()) };
+
+                if bytes_written < 0 {
+                    Err(VolatileMemoryError::IOError(std::io::Error::last_os_error()))
+                } else {
+                    Ok(bytes_written.try_into().unwrap())
+                }
+            }
+        }
+    }
+}
+
+pub(crate) trait IsUnixStream {
```

Comment:
Why a new trait and not just implementing `is_unix_stream` in StreamType?
I'd rename that method in just `is_unix`.

Reply:
sorry i am new to Rust so need some input on this. I added a trait because the type needs to be added in here https://github.com/rust-vmm/vhost-device/pull/706/files#diff-57d66eeb21fd6a88dd0baa6de0f6d922d1e00d2f1d593c69169278295f8b189aR59
for the type S . Can I do it without trait?

Reply:
Good point, I just realized stream is `S`, so what you did is right.

But instead of IsUnixStream, I'd implement a trait that tell us that the stream requires the "hybrid vsock" stuff.
So maybe something like `IsHybridVsock` ?